File: types/vault.ts
export type VaultState =
	| "open"
	| "withdrawable"
	| "trading"
	| "settled"
	| "closed";

export type VaultStatus = "active" | "inactive" | "all";

export interface Vault {
	id: string;
	state: VaultState;
	icon_url?: string;
	vault_name: string;
	summary?: string;
	annual_return: number;
	tvl_usd: number;
	max_drawdown?: number;
	start_time: string;
}

export interface VaultInfo {
	id: string;
	state: VaultState;
	icon_url?: string;
	vault_name: string;
	vault_type: string;
	vault_type_logo: string;
	blockchain: string;
	blockchain_logo: string;
	address: string;
	pool_id: string;
	summary?: string;
	description: string;
	annual_return: number;
	tvl_usd: number;
	max_drawdown: number;
	start_time: string;
	trade_per_month: number;
	decision_cycle: string;
}

export interface VaultsResponse {
	page: number;
	limit: number;
	offset: number;
	vaults: Vault[];
}

export type VaultResolution = "1d" | "1w" | "1m";
export type VaultCurrency = "usd" | "ada";

export interface VaultValuesRequest {
	resolution?: VaultResolution;
	currency?: VaultCurrency;
	start_time?: number;
	end_time?: number;
	count_back?: number;
}

export interface VaultValuesResponse {
	s: "ok" | "no_data";
	t: number[]; // timestamps
	c: number[]; // closing prices
	o?: number[]; // opening prices
	h?: number[]; // high prices
	l?: number[]; // low prices
	v?: number[]; // volume
}

export type PositionStatus = "open" | "closed";

export interface Position {
	pair: string;
	spend: number;
	value: number;
	profit: number; // profit percentage
	open_time: string;
	status: PositionStatus;
	close_time: string;
}

export interface PositionsResponse {
	total: number;
	page: number;
	limit: number;
	positions: Position[];
}

export interface PositionsRequest {
	status?: PositionStatus;
	page?: number;
	limit?: number;
	offset?: number;
}

export interface VaultStats {
	state: VaultState;
	tvl_usd: number;
	max_drawdown: number;
	trade_start_time: string;
	trade_end_time: string;
	start_value: number;
	current_value: number;
	return_percent: number;
	update_time: string;
	total_trades: number;
	winning_trades: number;
	losing_trades: number;
	win_rate: number;
	avg_profit_per_winning_trade_pct: number;
	avg_loss_per_losing_trade_pct: number;
	total_fees_paid: number;
	trade_per_month: number;
	decision_cycle: string;
	start_time: string;
}

export interface VaultEarning {
	vault_id: string;
	vault_name: string;
	vault_address: string;
	total_deposit: number;
	current_value: number;
	roi: number;
}

export interface UserVaultEarningInfoResponse {
	total_deposit: number;
	is_redeemed: boolean;
	profit_rate: number;
	min_deposit?: number;
	min_withdraw?: number;
}

export interface UserVaultEarningsRequest {
	wallet_address: string;
	limit?: number;
	offset?: number;
}

export interface UserVaultEarningsResponse {
	earnings: VaultEarning[];
	total: number;
	page: number;
	limit: number;
}

export type VaultTransactionType =
	| "deposit"
	| "withdrawal"
	| "claim"
	| "reinvest";

export interface VaultTransaction {
	id: string;
	vault_id: string;
	vault_name: string;
	wallet_address: string;
	action: VaultTransactionType;
	amount: number;
	token_id: string;
	token_symbol: string;
	txn: string;
	timestamp: number;
	status: "completed" | "pending" | "failed";
	fee: number;
}

export interface UserVaultTransactionsRequest {
	wallet_address: string;
	vault_id?: string;
	page?: number;
	limit?: number;
}

export interface UserVaultTransactionsResponse {
	transactions: VaultTransaction[];
	total: number;
	page: number;
	limit: number;
}

export interface VaultWithdrawRequest {
	vault_id: string;
	wallet_address: string;
	amount_ada?: number;
}

export interface VaultWithdrawResponse {
	status: "ok" | "invalid";
	tx_id: string | null;
	reason?: string;
}

File: app/schemas/vault.py
from typing import List, Literal, Optional

from pydantic import Field, field_validator

from app.core.config import settings
from app.schemas.my_base_model import CustomBaseModel


class VaultListItem(CustomBaseModel):
    """Vault list item for /vaults/{status} endpoint"""

    id: str = ""  # uuid
    state: str = ""  # open, trading, withdrawable, closed
    icon_url: Optional[str] = None
    vault_name: str = ""
    summary: Optional[str] = None
    address: str = ""
    pool_id: str = ""
    annual_return: float = 0.0
    tvl_usd: float = 0.0
    max_drawdown: float = 0.0
    start_time: Optional[int] = None


class VaultListResponse(CustomBaseModel):
    """Response model for vault list"""

    vaults: List[VaultListItem] = Field(default_factory=list)
    total: int = 0
    page: int = 1
    limit: int = 20


class VaultInfo(CustomBaseModel):
    """Vault info for /vaults/{id}/info endpoint"""

    id: str = ""  # uuid
    state: str = ""  # open, trading, withdrawable, closed
    icon_url: Optional[str] = None
    vault_name: str = ""
    vault_type: str = "Zenit Vault"
    vault_type_logo: str = settings.HOST + "/static/images/Zenit.png"
    blockchain: str = "polkadot"
    blockchain_logo: str = "https://cryptologos.cc/logos/polkadot-new-dot-logo.png"
    address: str = ""
    pool_id: str = ""
    summary: Optional[str] = None
    annual_return: float = 0.0
    tvl_usd: float = 0.0
    max_drawdown: float = 0.0
    start_time: Optional[int] = None
    trade_per_month: float = 0.0  # Average transactions per month f:.2
    decision_cycle: Optional[str] = None  # Decision cycle from trade strategy
    description: Optional[str] = None  # HTML text


class VaultValuesResponse(CustomBaseModel):
    """Vault values response for /vaults/{id}/values endpoint (TradingView format)"""

    s: str = "ok"  # Status code: ok, error, or no_data
    t: List[int] = Field(
        default_factory=list
    )  # Array of bar timestamps (Unix timestamp UTC)
    c: List[float] = Field(default_factory=list)  # Closing price


class VaultStats(CustomBaseModel):
    """Vault statistics for /vaults/{id}/stats endpoint"""

    state: str = ""  # open, trading, withdrawable, closed
    tvl_usd: float = 0.0
    max_drawdown: float = 0.0
    trade_start_time: Optional[int] = None
    trade_end_time: Optional[int] = None
    start_amount: float = 0.0
    current_amount: float = 0.0
    return_percent: float = 0.0
    total_trades: int = 0
    winning_trades: int = 0
    losing_trades: int = 0
    win_rate: float = 0.0
    avg_profit_per_winning_trade_pct: float = 0.0
    avg_loss_per_losing_trade_pct: float = 0.0
    total_fees_paid: float = 0.0
    decision_cycle: Optional[str] = None  # Decision cycle from trade strategy
    trade_per_month: float = 0.0  # Average transactions per month f:.2

    @field_validator("trade_per_month")
    def round_fields(cls, v: float) -> float:
        return round(v, 2)


class VaultPosition(CustomBaseModel):
    """Vault position for /vaults/{id}/positions endpoint"""

    pair: str = ""  # e.g., "ADA/USDM"
    spend: float = 0.0  # spend amount
    value: float = 0.0  # current value (return_amount if closed, estimated from current prices if open)
    open_time: int = 0  # position start_time
    close_time: Optional[int] = None  # position close_time
    status: str = ""  # "open" or "closed"
    profit: float = 0.0  # profit percentage: (value - spend) / spend * 100

    @field_validator("spend", "value", "profit")
    def round_fields(cls, v: float) -> float:
        return round(v, 2)


class VaultPositionsResponse(CustomBaseModel):
    """Response model for vault positions list"""

    total: int = 0
    page: int = 1
    limit: int = 20
    positions: List[VaultPosition] = Field(default_factory=list)

